src/main/java/hudson/plugins/mercurial/Cache.java

```diff
@@ -241,12 +251,19 @@ private synchronized ReentrantLock getLockForSlaveNode(String node) {
     /**
      * Hash a URL into a string that only contains characters that are safe as directory names.
      */
-    static String hashSource(String source, StandardUsernameCredentials credentials) {
+    static String hashSource(String source, StandardUsernameCredentials credentials, String masterCacheRoot) {
         if (!source.endsWith("/")) {
             source += "/";
         }
         Matcher m = Pattern.compile(".+[/]([^/:]+)(:\\d+)?[/]?").matcher(source);
-        String digestible = credentials == null ? source : source + '#' + credentials.getId();
+        String digestible = source;
+        if (credentials != null){
+            digestible +=  '#' + credentials.getId();
+        }
+        if (masterCacheRoot != null){
+            digestible += "#" + masterCacheRoot.replaceAll(File.pathSeparator, "_");
```

Comment:
OK…irrelevant since a given master is either running on Windows or it is running on Unix-ish, and that is very unlikely to change between restarts. But harmless I suppose.